request-manager/graph/resolver.go

```diff
@@ -142,95 +144,65 @@ func (o *resolver) BuildJobChain(jobArgs map[string]interface{}) (*proto.JobChai
-		}
-	}
-	for _, arg := range seq.Args.Static {
-		if _, ok := jobArgs[*arg.Name]; !ok {
-			jobArgs[*arg.Name] = *arg.Default
-		}
-	}
-
-	/* Build request graph based on sequence graph. */
-	seqGraph, ok := o.SequenceGraphs[sequenceName]
+	// Build request graph based on sequence graph.
+	seqGraph, ok := o.sequenceGraphs[sequenceName]
 	if !ok {
+		// Graph checks should prevent this from happening.
+		// If this error is thrown, there's a bug in the code.
 		return nil, fmt.Errorf("cannot find sequence graph for sequence: %s", sequenceName)
 	}
 
-	reqGraph, err := o.newReqGraph(cfg.wrapperName, jobArgs)
+	reqGraph, err := o.newReqGraph(cfg.graphName, jobArgs)
 	if err != nil {
 		return nil, err
 	}
 
-	// Traverse sequence graph in topological order. Each sequence graph node corresponds to either a single
-	// job node, or an entire subgraph of nodes. Build the subgraph and insert it into the larger sequence
-	// graph.
+	// Traverse sequence graph in topological order. Each sequence graph node
+	// corresponds to either a single job node, or an entire subgraph of nodes.
+	// Build the subgraph and insert it into the larger sequence graph.
 	idMap := map[string]*Graph{} // sequence graph node id --> corresponding subgraph
 	for _, seqGraphNode := range seqGraph.Order {
-		n := seqGraphNode.Spec
-
-		// Find out how many times this node has to be repeated
-		elements, lists, err := o.getIterators(n, jobArgs)
+		nodeSpec := seqGraphNode.Spec
+
+		// 1. Do sequence expansion.
+		// Find out how many times this node has to be repeated.
+		// If this node has a `each:` field, then lists[i] and
+		// elements[i] describe the ith `list:element`. Specifically,
+		// lists[i] is the actual list value, not just the name of the
+		// job arg.
+		elements, lists, err := o.getIterators(nodeSpec, jobArgs)
 		if err != nil {
-			return nil, fmt.Errorf("in seq %s, node %s: invalid 'each:' %s", sequenceName, n.Name, err)
+			return nil, fmt.Errorf("in seq %s, node %s: invalid 'each:' %s", sequenceName, nodeSpec.Name, err)
 		}
 
-		// All the graphs that make up this sequence graph node's subgraph
+		// All the graphs that make up this sequence graph node's subgraph,
+		// one for each time the node is repeated
 		components := []*Graph{}
 
 		// If no repetition is needed, this loop will only execute once
+		// with a dummy `each:` entry `[""]:""`.
 		for i, _ := range lists[0] {
-			// Copy the required args into a separate args map here.
-			// Do the necessary remapping here.
-			jobArgsCopy, err := remapNodeArgs(n, jobArgs)
+			// Copy the required args into a separate args map here
+			// and do the necessary remapping
+			jobArgsCopy, err := remapNodeArgs(nodeSpec, jobArgs)
 			if err != nil {
 				return nil, err
 			}
 
-			// Add the element to the node args unless there is none for this node
+			// Add elements to the node args unless there are none
+			// for this node
 			for j, elt := range elements {
 				if elt != "" {
```

Comment:
This block is unclear but very important. What's really happening here is: Given "each: foos:foo", we magically (good magic!) create jobArgs[foo]=jobArgs[foos][i]. This also means that `if elt != ""` is confusing because does it mean jobArgs[foos] can't have any empty string values? I.e. we shouldn't make "" a magical value here.